ansible/roles/csm.rr.ceph_zoning/files/ceph_zoning.py

```diff
@@ -0,0 +1,194 @@
+#
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+
+#This script is used to apply CEPH zoning
+#The objective of CEPH zoning is to make sure data gets replicated at rack level, so there would not be data loss incase of a rack failure
+
+import json
+import subprocess
+import re
+import sys
+import logging
+import base64
+
+# Set up logger
+logger = logging.getLogger("CephZoning")
+logger.setLevel(logging.DEBUG)
+ch = logging.StreamHandler()
+ch.setLevel(logging.DEBUG)
+formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
+ch.setFormatter(formatter)
+logger.addHandler(ch)
+
+def get_ceph_zone_prefix():
+    # Run kubectl command and capture JSON output
+    namespace = "loftsman"
+    secret_name = "site-init"
+
+    kubectl_cmd = ["kubectl", "-n", namespace, "get", "secret", secret_name, "-o", "json"]
+    kubectl_output = subprocess.run(kubectl_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True, check=True)
+
+    # Parse JSON output
+    secret_data = json.loads(kubectl_output.stdout)
+
+    # Extract and decode the base64 data
+    encoded_yaml = secret_data["data"]["customizations.yaml"]
+    decoded_yaml = base64.b64decode(encoded_yaml).decode("utf-8")
+
+    # Write the yaml output to a file
+    output_file = "/tmp/customizations.yaml"
+    with open(output_file, "w") as f:
+        f.write(decoded_yaml)
+
+    # Define the key path
+    output_file = "/tmp/customizations1.yaml"
+    ceph_key_path = "spec.kubernetes.services.ceph_zone_prefix"
+
+    # Run yq command to extract the value
+    ceph_yq_cmd = ["yq", "r", output_file, ceph_key_path]
+    ceph_zone = subprocess.run(ceph_yq_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True, check=True)
+
+    # Extract and clean the output
+    ceph_zone_prefix = ceph_zone.stdout.strip()
+    return ceph_zone_prefix
+
+def run_command(command):
+    """Helper function to run a command and return the result."""
+    logger.info(f"Running command: {command}")
+    try:
+        result = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True, shell=True, check=True)
+    except subprocess.CalledProcessError as e:
+        raise ValueError(f"Command {command} errored out with : {e.stderr}")
+    return result.stdout
+
+def create_and_map_racks(positions_dict):
```

Comment:
Please add docstrings and type annotate function arguments and return types. For example: https://github.com/Cray-HPE/bos/blob/564dab11047ab0f198b3e4f0de39b7568d9322d5/src/bos/server/redis_db_utils/dbwrapper.py#L174-L185

Reply:
We at least need to add type annotations before we merge the code. As it is now, it is difficult to review the code because it's not clear what types the method arguments are supposed to be, and it is a lot of work to trace through the code to figure it out.